Implement analysis server closing labels

DanTup opened the pull request:
@bwilkerson I've coded a basic implementation for #30285 with some tests. I'm creating this PR as somewhere to get feedback/review but it'll likely need some work before it's done.

 I don't know if you'd rather wait until they're done before reviewing or review in small chunks (similarly, I don't know whether you'd rather merge multiple PRs or just keep this until it's complete).

I haven't done a thorough test against real world code (valid nor invalid) yet so this is mostly just "the happy path", but I hope to do some proper testing later this week. I don't know how thorough you'd consider these tests but if any test cases/edge cases jump out, lmk and I'll add them.

We might also want to discuss how big an issue bad handling of errors (like #30370) is.

We'll also need to increase the API version when/after landing this so that clients know that this is available too.

DanTup committed:
Add empty closing labels computer/visitor and tests.

See #30285

DanTup committed:
Add initial implementation and test for closing labels.

See #30285

DanTup committed:
Add various tests for analyzer closing labels.

See #30285

DanTup committed:
Fail tests if errors occur.

Without this the notifications will never come through and the test will time out. This causes them to fail right away on errors.

See #30285

DanTup committed:
Add test and fix for crash that occurred during testing.

See #30285

bwilkerson commented on pkg/analysis_server/lib/src/computer/computer_closingLabels.dart:
```diff
@@ -0,0 +1,85 @@
+  @override
+  Object visitInstanceCreationExpression(InstanceCreationExpression node) {
+    if (_spansManyLines(node)) {
+      // TODO(dantup) Is it right to read constructorName to get nameless constructors types?
```
It won't necessarily give you the actual name of the class of object that will be returned (in the case of a factory constructor), but it will match what's in the code, so I'd say yes.

bwilkerson commented on pkg/analysis_server/lib/src/computer/computer_closingLabels.dart:
```diff
@@ -0,0 +1,85 @@
+    if (_spansManyLines(node)) {
+      // TODO(dantup) Is it right to read constructorName to get nameless constructors types?
+      final name = node.constructorName.type.name;
+      final label = name is PrefixedIdentifier ? name.prefix.name : name.name;
```
I don't think we want to display just the name of the prefix. We can include the prefix if you want, but for a prefixed identifier it's the `identifier` that matters. In other words, I think this should be either
```
final label = name.name;
```
or
```
final label = name is PrefixedIdentifier ? name.identifier.name : name.name;
```

DanTup replied:
Using just `name.name` results in `a.Future.delayed` becoming `a.Future`, but I think either `Future` or `Future.delayed` would make sense here.

Confusingly, a non-prefixed named constructor behaves differently (returning `Class.fromThing`) which is what I'd actually been testing when I added this code.

I'll add tests for prefixed imports and find a way to handle both of these cases in a good way.

DanTup replied:
Ok, I see why I'm confused.. For `a.Future.delayed`, the `constructorName.type` seems to be `a.Future` (sounds reasonable) and `constructorName.name` is `delayed` (also seems reasonable).

However, for `Class.fromThing`, `constructorName.type` seems to include `fromThing` and `constructorName.name` is null. This isn't what I'd expect; I'd expect `constructorName.type` to be `Class` and `constructorName.name` to be `fromThing`.

<img width="301" alt="screen shot 2017-08-09 at 18 09 13" src="https://user-images.githubusercontent.com/1078012/29134209-e07552fa-7d2d-11e7-92b2-92a19cd40cf7.png">

What's the reason for this (or what am I getting confused by)?

Btw - doing `node.constructorName.toString()` seems to give a reasonable output here, but it feels slightly weird to be calling `toString` on the AST node for this. WDYT?

bwilkerson replied:
When the parser sees `new a.b` it can't tell whether `a` is a prefix used to access a class named `b` or whether there is a class named `a` with a constructor named `b`. It guesses one and is sometimes wrong. I had plans, at one point, to fix the AST after resolution (when we'd know how we should have parsed it) but that hasn't happened yet.

Yes, it is kind of weird to call `toString`. We use `toString` strictly as a debugging aid so it should never be called from production code. If we need to we can consider adding a `displayName` getter (or some such) to ConstructorName.

DanTup replied:
Just after posting I realised it can't tell at this point. I presume fixing it up after resolution won't help here, since I don't think it's being resolved.

Thanks for confirming about toString, I'll come back to this!

bwilkerson commented on pkg/analysis_server/lib/src/computer/computer_closingLabels.dart:
```diff
@@ -0,0 +1,85 @@
+
+  @override
+  Object visitMethodInvocation(MethodInvocation node) {
+    if (node.argumentList != null && _spansManyLines(node.argumentList)) {
```
I'm curious as to why this isn't `_spansManyLines(node)`? Seems like it matches the contructor case better that way.

DanTup replied:
Good question! I thought it was maybe a hangover from when I only excluded things on the same line, so args pushed onto the next line (but on a single line) would otherwise result in a label; however I can't actually reproduce that now (the arg list seems to start on the same line with the paren). I'll change to `node`.

bwilkerson commented on pkg/analysis_server/lib/src/computer/computer_closingLabels.dart:
```diff
@@ -0,0 +1,85 @@
+
+  bool _spansManyLines(AstNode node) {
+    final start = computer._lineInfo.getLocation(node.offset);
+    // TODO(dantup) Is this valid? We want the last character of the node, not the start of the next?
```
I believe so, yes. But note that `end` doesn't return the offset of the next token, only the next character (which could be whitespace, including an end-of-line marker).

bwilkerson commented on pkg/analysis_server/lib/src/computer/computer_closingLabels.dart:
```diff
@@ -0,0 +1,85 @@
+    // TODO(dantup) Is this valid? We want the last character of the node, not the start of the next?
+    final end = computer._lineInfo.getLocation(node.end - 1);
+
+    return (start.lineNumber - end.lineNumber).abs() > 1;
```
If you reverse the order of the subtraction you won't need `.abs()`.

bwilkerson commented on pkg/analysis_server/test/analysis/notification_closingLabels_test.dart:
```diff
@@ -0,0 +1,311 @@
+}
+
+@reflectiveTest
+class _AnalysisNotificationClosingLabelsTest extends AbstractAnalysisTest {
```
You should include some tests in which some of the type names are prefixed (which will require, iirc, using prefixed named constructors like `new p.C.c(...)`).

That said, I would break these tests up a bit. Having a couple of tests that create a server, subscribe for and listen for notifications, and compare the result (what you have here) is good. But I'd move most of the tests into a set of unit tests for `DartUnitClosingLabelsComputer` because there's less overhead, so the tests will run faster. Not a problem if you don't agree or don't want to do that.

DanTup committed:
Remove unnecessary abs().

See #30285

DanTup committed:
Remove question that has been answered (yes).

See #30285

DanTup committed:
Remove unnecessary .argumentList on spansManyLines check.

See #30285

DanTup committed:
Add test for prefixed import and constructor names.

The behaviour seems like what we want, though need to confirm if this is the best way to do it.

See #30285

DanTup committed:
Add additional prefixed tests and another TODO.

See #30285

DanTup commented:
Ok, I've pushed some changes to address most of the comments here. Outstanding from the comments above are:

1. Stop using toStrings to get labels (I'll build what I need from the other properties for now and we can discuss whether it's worth making `displayNames` from that logic later)
2. Make `prefix.Class.constructor` and `prefix.Class.staticMethod` consistent (currently the constructor gets the full string and the static method only gets the method name)
3. Convert most of the tests to bypass the service and just test the computer directly

DanTup commented:
Strange failure on Travis just before the last changeset I pushed; seems to be this:

```text
Chrome 60.0.3112 (Linux 0.0.0)  corelib/iterable_fold_test_02_multi FAILED
	Error: Type '(dynamic) -> dynamic' is not a subtype of type '(int) -> bool'
	    at finish (test/browser/language_tests.js:768:59)
	    at Context.<anonymous> (test/browser/language_tests.js:791:11)
```

The error message doesn't look like something that could be flakey, but it also doesn't sound like one of my tests (these don't get run in the browser, right?). Any ideas? (it's building again from the last changeset, so it's possible it'll go green; though that doesn't explain what happened).

bwilkerson commented:
The test that's failing is a shared test that cannot be impacted by the work you're doing. It's either flaky or someone committed a change that broke the test and your run got scheduled between that and the commit to fix it. Either way, nothing you need to worry about.

DanTup commented:
Ah; I thought it was building just my PR branch in isolation (which only I'm committing to), but looks like it's merging first. Makes sense!

DanTup committed:
Fix method invocations to show full target/name.

See #30285

DanTup committed:
Fix constructor invocations to not use toString().

See #30285

DanTup committed:
Convert most closing label tests to unit tests.

See #30285

DanTup commented:
@bwilkerson I think I've addressed all of the existing comments now, including moving most tests to unit tests.

I still need to do some more thorough testing with some large codebases - I'll add tests for any errors I find and also make a list of any places where invalid code returns unexpected results. LMK if there's anything else you want me to do or any further changes. Thanks!

bwilkerson commented:
Should I merge now (my preference), or were you planning on writing more tests first?

DanTup commented:
@bwilkerson I might add more tests if anything comes out of the testing, but if you'd prefer to merge it now (and are happy it's in a reasonable state) I can send another PR for any further changes.

We'll need to update the server version too at some point; though maybe better after I've done a little more testing.

DanTup commented:
@bwilkerson 

> I'm curious as to why this isn't _spansManyLines(node)? Seems like it matches the contructor case better that way.

I found the reason I'd done this... Given this code:

```dart
return _closingLabelsByEndLine.values
    .where((l) => l.any((cl) => cl.spannedLines >= 2))
    .expand((cls) => cls)
    .map((clwlc) => clwlc.label)
    .toList();
```

The ast node for the `map` method invocation has an offset that starts at `_closingLabels...` (incorrectly resulting in a label) whereas the argument list starts where the args are.

So, I'm going to put this back for now - I'll send a PR soon with this and some other fixes; and if you think it's the wrong fix we can discuss there (since this one is merged).

I'll also test similar for constructors to see if it has the same issue if prefix.Class.ctor gets split.
